second attempt to add sd-jwt vcdm (#147)
openid/oid4vc-haip: openid4vc-high-assurance-interoperability-profile-1_0.md

A Credential Format Profile for Credentials complying with IETF SD-JWT VCs [@!I-D.ietf-oauth-sd-jwt-vc] is defined in Annex A.3 of [@!OIDF.OID4VCI] and Annex A.4 of [@!OIDF.OID4VP].

## SD-JWT VC Data Model (SD-JWT VCDM)

> SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0], while enabling selective disclosure.

Reviewer: Please change this sentence. While admittedly not strictly untrue, to most readers, this paragraph will only perpetuate the false claim that W3C VCDM does not support selective disclosure, which has been spread at various conferences and communities in the last years. The claim that SD-JWT VCDM is needed to do SIMPLE selective disclosure with W3C VCDM is equally untrue, as you can also just usw plain SD-JWT if you really want. The introduction should explain this to avoid risk of further confusion and deception.

> SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0], while enabling selective disclosure.

Reviewer 2: This, of course, doesn't strictly speak to the false cry of perpetuating claims that W3C VCDM does not support selective disclosure but might be more relatable to many readers. 
```suggestion
SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0], while enabling a consistent approach to selective disclosure.
```

Reviewer: > false cry

I don't think it's helpful to bring emotions into this thread. There is no "cry", and it is not "false" to simply observe that SD-JWT (VC(DM)) proponents have made false claims about selective disclosure in W3C VCDM.

Reviewer: To try and be productive, I don't think this suggested change is helpful, since it still doesn't explain why SD-JWT VCDM is needed. If you want to enable "a consistent approach to selective disclosure", as this suggestion says, you can also just use plain SD-JWT with VCDM directly, without the SD-JWT VC extensions.

Therefore I still think this introduction needs to explain somehow what this work adds to what's already out there.

SD-JWT VCDM credentials contain a data structure that can be processed using a JSON-LD processor after applying SD-JWT processing.

When IETF SD-JWT VC is mentioned in this specification, SD-JWT VCDM defined in this section MAY be used.

> Implementaters of SD-JWT VCDM MUST use valid values for both `vct` Claim defined in IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] and `type` proprty defined in W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].

Reviewer 2: ```suggestion
Implementers of SD-JWT VCDM MUST use valid values for both `vct` Claim defined in IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] and `type` proprty defined in W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].
```

> Implementaters of SD-JWT VCDM MUST use valid values for both `vct` Claim defined in IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] and `type` proprty defined in W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].

Reviewer 3: ```suggestion
Implementaters of SD-JWT VCDM MUST use valid values for both `vct` Claim defined in IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] and `type` property defined in W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].
```

For backward compatibility with JWT processors, the following registered JWT claims MUST be used, instead of their respective counterpart properties in W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0]:

> * To represent the validity period of SD-JWT VCDM (i.e., cryptographic signature), `exp`/`iat` Claims encoded as a UNIX timestamp (NumericDate) MUST be used, and not `expirationDate` and `issuanceDate` properties.

Reviewer: What about `validFrom` and `validTo` in VCDM 2.0 ?

Reviewer 4: As implementing acts are published, let's please refer to the data model in the implementing act:
https://eur-lex.europa.eu/eli/reg_impl/2024/2977/oj

Reviewer 5: > What about `validFrom` and `validTo` in VCDM 2.0 ?

@peacekeeper this is also the approach taken by W3C VCWG COSE-JOSE to use JOSE claims for the validity of the signatures, and use other claims to express document validity. 

@alenhorvat Would you agree that this is also sufficient for the data model defined by the IAs? IMO, `expiry_date` etc. are claims related to document validity and evaluated by the business logic that knows the data model.

Author: ```suggestion
* To represent the validity period of SD-JWT VCDM (i.e., cryptographic signature), `exp`/`iat` Claims encoded as a UNIX timestamp (NumericDate) MUST be used, and not `expirationDate` and `issuanceDate` properties defined in [@!W3C.VCDM1.1], `validFrom` and `validTo` properties defined in [@!W3C.VCDM2.0].
```

Author: good catch, Markus, thank you!

> * To represent the validity period of SD-JWT VCDM (i.e., cryptographic signature), `exp`/`iat` Claims encoded as a UNIX timestamp (NumericDate) MUST be used, and not `expirationDate` and `issuanceDate` properties defined in [@!W3C.VCDM1.1], `validFrom` and `validTo` properties defined in [@!W3C.VCDM2.0].

Reviewer 3: ```suggestion
* To represent the validity period of SD-JWT VCDM (i.e., cryptographic signature), `exp`/`nbf` Claims encoded as a UNIX timestamp (NumericDate) MUST be used, and not `expirationDate` and `issuanceDate` properties defined in [@!W3C.VCDM1.1], `validFrom` and `validTo` properties defined in [@!W3C.VCDM2.0].
```

Reviewer 3: See:

- https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.5
- https://www.w3.org/TR/vc-data-model-2.0/#validity-period

`iat` is the time the token is issued, not the time at which the token is to be considered valid.

It's conceivable that a mismatch here would prevent use cases where I issue a credential now, that becomes active a day from now.

You could also say that for simplicity it is not possible to issue sd-jwt-vc with validity periods which exist in the future, and leave the current text, and resolve this comment.

Reviewer 2: related https://github.com/eu-digital-identity-wallet/eudi-doc-architecture-and-reference-framework/pull/160#discussion_r1952196890

* `iss` Claim MUST represent the identifier of the `issuer` property, i.e., the `issuer` property value if `issuer` is a String, or the `id` property of the `issuer` object if `issuer` is an object. `issuer` property MUST be ignored if present.
* `status` Claim MUST represent `credentialStatus` property. `credentialStatus` property MUST be ignored if present.

> * `sub` Claim MUST represent the `id` property of `credentialSubject` property. `credentialSubject` property MUST be ignored if present.

Reviewer 3: ```suggestion
* `sub` Claim MUST represent the `id` property of `credentialSubject` property. `credentialSubject` property MUST be ignored if present. There is no support for the case where `credentialSubject` is an array.
```

IETF SD-JWT VC is extended with the following claims:

> * `vcdm`: OPTIONAL. Contains properties defined in [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0] that are not represented by their counterpart JWT Claims as defined above.

Reviewer 3: Don't do this.

Just let the claimset be defined as a JWT claimset.
And allow any properties that are not understood to be ignored.
Do not introduce a nested object, it breaks compatibility with the TR defined here:

https://www.w3.org/TR/vc-jose-cose/#securing-with-sd-jwt

Reviewer: Agree with @OR13 , inserting one data model as a nested object inside another data model will lead to interop problems.

The following outlines a suggested non-normative processing steps for SD-JWT VCDM:

1. SD-JWT VC Processing:

- A receiver (holder or verifier) of an SD-JWT VCDM applies the processing rules outlined in Section 4 of [@!I-D.ietf-oauth-sd-jwt-vc], including verifying signatures, validity periods, status information, etc.

> - If the `vct` value is associated with any SD-JWT VC Type Metadata, schema validation of the entire SD-JWT VCDM is performed, including the nested `vcdm` claim.

Reviewer 3: Just say that when `vct` is present it overrides any application profile type specific processing that might have been implied by `type` as described in https://www.w3.org/TR/vc-data-model-2.0/#types

2. Business Logic Processing:

- Once the SD-JWT VC is verified and trusted by the SD-JWT VC processor, and if the `vcdm` claim is present, the receiver extracts the VCDM (1.1 or 2.0) object from the `vcdm` claim and uses this for the business logic object. If the `vcdm` claim is not present, the entire SD-JWT VC is considered to represent the business logic object.
- The business logic object is then passed on for further use case-specific processing and validation. The business logic assumes that all security-critical functions (e.g., signature verification, trusted issuer) have already been performed during the previous step. Additional schema validation is applied if provided in the `vcdm` claim, e.g., to support SHACL schemas. Note that while a `vct` claim is required, SD-JWT VC type metadata resolution and related schema validation is optional in certain cases.

The following is a non-normative example of an unsecured payload of an SD-JWT VCDM, that is built using the example of unsecured payload in Section 3.3 of [@!I-D.ietf-oauth-sd-jwt-vc]:

```json
{
  "vct": "https://credentials.example.com/identity_credential",

>   "vcdm": {    //W3C VCDM 2.0 compliant claims

Reviewer: What does this mean for:

> IANA reserved `vc` for w3c vcdm 1.1

    "@context": ["https://www.w3.org/ns/credentials/v2"],
    "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],
    "credentialSubject": {
      "given_name": "John",

>       "family_name": "Doe",

Reviewer 3: these terms are not defined in https://www.w3.org/ns/credentials/v2, so you either need to add another context to the top level, or leave them defined like this to highlight that the RDF part of JSON-LD rarely works.

      "email": "johndoe@example.com",
      "phone_number": "+1-202-555-0101",
      "address": {
        "street_address": "123 Main St",
        "locality": "Anytown",
        "region": "Anystate",
        "country": "US"
      },
      "birthdate": "1940-01-01",
      "is_over_18": true,
      "is_over_21": true,
      "is_over_65": true
    }
  }
}
```

> The following is a non-normative example of how an unsecured payload of an SD-JWT VCDM above can be used for the SD-JWT:

Reviewer 2: Still not sure this sentience makes sense but the "an"s here weren't helping. 
```suggestion
The following is a non-normative example of how the unsecured payload of the SD-JWT VCDM above can be used for the SD-JWT:
```

Reviewer 3: I think this sentence is trying to say:

The following is a non-normative example of how to secure "application/vc" with "application/dc+sd-jwt":

```json
{
  "_sd": [

>     "vwUhdFvpylx9Sqi2YNBV1dfVK6lCbhXvkH0nThfKFT0"

Reviewer 2: I realize it's just a "a non-normative example" but to my naïve little mind this implies that the `vcdm` claim is the only thing made selectively disclosable. And that doesn't make much sense to me.

## SD-JWT VC Data Model (SD-JWT VCDM)

SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0], while enabling selective disclosure.

SD-JWT VCDM credentials contain a data structure that can be processed using a JSON-LD processor after applying SD-JWT processing.

When IETF SD-JWT VC is mentioned in this specification, SD-JWT VCDM defined in this section MAY be used.

> Implementaters of SD-JWT VCDM MUST use valid values for both `vct` Claim defined in IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] and `type` proprty defined in W3C VCDM [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].

Reviewer: Do they have to match or not? What happens if one client processes the `vct` claim, and another processes the `type` claim, and they arrive at different conclusions? To me it sounds like a huge source of ambiguity and potential problems if a data model contains basically the same concept twice ("type of a Verifiable Credential"). Especially if not only the values can be different, but also the way how the values are being expressed is different.

Saying that "it's the responsibility of the issuer" is a bit naïve I think...

A Credential Format Profile for Credentials complying with IETF SD-JWT VCs [@!I-D.ietf-oauth-sd-jwt-vc] is defined in Annex A.3 of [@!OIDF.OID4VCI] and Annex A.4 of [@!OIDF.OID4VP].

## SD-JWT VC Data Model (SD-JWT VCDM)

> SD-JWT VCDM is a data model that follows IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc], but allows the usage of [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0]. When IETF SD-JWT VC is mentioned in this specification, SD-JWT VCDM define in this section MAY be used.

Author: need to add more detailed explanation that sd-jwt processor is used to process sd-jwt to get the payload, and json-ld processor is only applied to the vcdm part of the payload. also clarify that sd-jwt before processing (the one that includes _sd etc) should not be put into a json-ld processor.

Author: @awoie to propose a spcific text on processing :)

Reviewer 5: Done, see here https://github.com/openid/oid4vc-haip/pull/147#discussion_r1922554997

Author: thank you!

Reviewer 4: Proposal to resolve the problem
https://github.com/openid/oid4vc-haip/pull/147#discussion_r1932025108

Author: ok, let's continue the discussion there

> SD-JWT VCDM is a data model that follows IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc], but allows the usage of [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0]. When IETF SD-JWT VC is mentioned in this specification, SD-JWT VCDM define in this section MAY be used.

Author: ```suggestion
SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0], while enabling selective disclosure.

SD-JWT VCDM credentials contain a data structure that can be processed using a JSON-LD processor after applying SD-JWT processing.

When IETF SD-JWT VC is mentioned in this specification, SD-JWT VCDM defined in this section MAY be used.
```

> For backward compatibility with JWT processors, the following registered JWT claims MUST be used, instead of their respective counterpart properties in [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0]:

Author: need to add a paragraph on vct claim and type property

Reviewer 4: The data model should not be impacted by DCQL. If it is, there's an issue in the design and should be resolved.

> * `exp` Claim MUST represent the `expirationDate` property, encoded as a UNIX timestamp (NumericDate).

Author: ```suggestion
* `expirationDate` property MUST NOT be used. It MUST be ignored if present. It is represented by `exp` Claim encoded as a UNIX timestamp (NumericDate).
```

Author: this is not a transformation logic

Reviewer 4: Official claim is called expiry_date
https://eur-lex.europa.eu/eli/reg_impl/2024/2977/oj

> * `iss` Claim MUST represent the `issuer` property.

Author: ```suggestion
* `issuer` property MUST NOT be used. It MUST be ignored if present. It is represented by `iss` Claim.
```

leaving it to the implementations how to turn issuer into iss when issuer is an object

> * `iss` Claim MUST represent the `issuer` property.

Reviewer 4: iss cannot represent the issuer claim as issuer can be a string or an object, while iss is a string with well-defined syntax (especially for jwt processors), that go and resolve the Authorization Server configuration from where they figure out the Authorization Server JWT signing key.

is 'iss' redefined in this document?

> * `iss` Claim MUST represent the `issuer` property.

Reviewer 5: I would keep the original line for the following reason.

The SD-JWT VC processing is first since it deals with security-critical functions. `iss` is used by SD-JWT VC to validate the public verification key of the issuer (as per SD-JWT VC). To make sure self-declared und unverifiable values are not permitted, the linkage between the `iss` and `vcdm.issuer` (if String) /`vcdm.issuer.id` (if `vcdm.issuer` is an object) has to be kept for integrity reasons since the business logic (if `vcdm` claim is included) would rely on the SD-JWT VC verification logic that also checks the issuer.

> * `iss` Claim MUST represent the `issuer` property.

Reviewer 4: There should not be vcdm.issuer

Is "iss" redefined in this document? If not, which definition is being followed? The one of OIDC? Any other?
RFC https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.1 is not defining what "iss" should be.

OIDC definition:
iss
REQUIRED. Issuer Identifier for the Issuer of the response. The iss value is a case-sensitive URL using the https scheme that contains scheme, host, and optionally, port number and path components and no query or fragment components.

Which definition is followed? (any other?)

* `iat` Claim MUST represent `issuanceDate` property, encoded as a UNIX timestamp (NumericDate).

> * `status` Claim MUST represent `credentialStatus` property.

Reviewer 4: Do you confirm that status and credential status have the same governance model? So, it contains a "type" property that defines which revocation/suspension mechanism is used?

Reviewer 4: Kind of is not an answer. It should be clearly expressed how different status methods can be used, such as CRL, OCSP.

> * `schema` Claim MUST represent the `credentialSchema` property.

Reviewer 4: Same question as for status; Note: ELM is using 2: JSON schema and SHACL schema.

Author: kind of, I guess.

Author: ```suggestion
```
removing since `schema` property belongs in the vcdm container

* `sub` Claim MUST represent the `id` property of `credentialSubject` property.

IETF SD-JWT VC is extended with the following claims:

* `vcdm`: OPTIONAL. Contains properties defined in [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0] that are not represented by their counterpart JWT Claims as defined above.

> The following is a non-normative example of an unsecured payload of an SD-JWT VCDM, that is built using the example of unsecured payload in Section 3.3 of [@!I-D.ietf-oauth-sd-jwt-vc]:

Author: expectation is for all user claims being included in vcdm, not outside of it. it makes more sense to define a static vct that is used when vcdm claim is present. and that static vct can have a flag that json-ld processing applies to vcdm. but this might be too complex

Reviewer 4: How is sd-jwt vc defining "namespaces"/"vocabularies"?
Namely, all the issues raised here are due to lack of it.

Reviewer: > How is sd-jwt vc defining "namespaces"/"vocabularies"?

sd-jwt vc has invented its own way of doing this, called "JWT VC Type Metadata"

Reviewer 4: Interesting. How do you (unambiguously) represent a claim from:
a) IANA (any IANA registered claim, not only JWT registered claim)
b) claim registered in OID? (used by x509 and mDOC/mdl)
c) public vocabulary, such as ELM
d) private vocabulary

Reviewer 6: With the current proposal, everything outside the `vcdm` claim would be IANA/standard JWT claims; everything inside would be the same as in W3C VCDM.

Reviewer 4: @awoie and @danielfett , I would kindly ask you to answer the question:

How do you (unambiguously) represent a claim from:
a) IANA (any IANA registered claim, not only JWT registered claim)
b) claim registered in OID? (used by x509 and mDOC/mdl)
c) public vocabulary, such as ELM
d) private vocabulary

Note: JSON-LD is not mentioned anywhere. In my view, as expressed several times, JSON-LD is not posing any issues at all, if positioned correctly. (as JSON-LD extends JSON with additional properties, it's very easy to model the solution with clear boundaries)

```json
{
  "vct": "https://credentials.example.com/identity_credential",
  //W3C VCDM 2.0 compliant claims

>   "vcdm": {

Reviewer 4: Why?

Author: resolving since discussion is continuing in another thread

    "@context": ["https://www.w3.org/ns/credentials/v2"],

>     "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],

Reviewer 4: should be top level and relationship with vct must be explained

Reviewer 6: There are various options, see above, but it seems that only one is reasonable.

Reviewer 7: Please correct me if I am wrong:
SD-JWT VC encodes credentials using a JSON-based structure, representing claims as key-value pairs. The focus is on simplicity and compatibility with existing JSON and JWT ecosystems.
Unlike RDF (W3C VC (DM)), which supports rich semantic relationships and graph-based representations, SD-JWT VC does not inherently provide mechanisms for linking data with semantic meaning. 

My question: How can the Trusted Schema Registry be integrated properly?
On which "specifications" or documentation is this discussion based on?

A Credential Format Profile for Credentials complying with IETF SD-JWT VCs [@!I-D.ietf-oauth-sd-jwt-vc] is defined in Annex A.3 of [@!OIDF.OID4VCI] and Annex A.4 of [@!OIDF.OID4VP].

## SD-JWT VC Data Model (SD-JWT VCDM)

> SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0], while enabling selective disclosure.

Reviewer: I think it's time to stop insinuating that VCDM doesn't support selective disclosure. This has always been a false claim by the SD-JWT (VC(DM)) community.

Considering this, what is the rationale for a hybrid format like SD-JWT VCDM in the first place? What can you do with this that you can't already do with VCs? Previous attempts like this ([VC-JWT](https://www.w3.org/TR/vc-data-model/#json-web-token)) have not been very successful and have created huge interoperability problems, so why repeat the same mistake?

If "Simple is a feature" - as we have often heard from the SD-JWT community in the last few years - then complex hybrid models like this should be avoided.

Reviewer 5: @peacekeeper Honestly, the W3C VC WG decided that the VCDM has to be a valid JSON-LD object in compact form long time ago. While there is probably some benefit of doing this, a lot of people just don't see that benefit. For them, JSON-only is good enough and anything related to RDF/JSON-LD and so on would require a massive amount of extra effort without a real benefit to them. It makes no sense to talk about the benefits here.

I'll argue that VC-JWT was very successful but had some interop issues, correct.

Reviewer 6: Another rationale is that this approach would allow to use the exact same selective disclosure mechanism for all credentials, whether they make use of W3C VCDM (JSON-LD) data structures or not.

Reviewer: @awoie 

> VCDM has to be a valid JSON-LD object

> JSON-only is good enough

> It is not possible to convince either side of this conversation.

Even if we accept that some want to use JSON-LD and some don't want to use it, and even if accept that (sadly) a compromise within the W3C VC WG was not possible, this still doesn't explain why the world needs SD-JWT VCDM.

Let those who prefer JSON use JWTs (with extensions like SD-*, type metadata, token status list, etc.), and let those who prefer JSON-LD use VCDM. Both formats are complex, and they are diverging more and more. Creating a hybrid feels like multiplying the complexities of both.

Didn't you and I have a Zoom call a few months ago where we agreed that those shouldn't be mixed, or am I misremembering that (in that case, sorry)?

Reviewer: @danielfett 

> this approach would allow to use the exact same selective disclosure mechanism

This doesn't explain a need for SD-JWT VCDM either, as it's also possible to use VCDM as an SD-JWT payload directly (see examples in the VCDM 2.0 specification).

Reviewer 5: > Let those who prefer JSON use JWTs (with extensions like SD-*, type metadata, token status list, etc.), and let those who prefer JSON-LD use VCDM. Both formats are complex, and they are diverging more and more. Creating a hybrid feels like multiplying the complexities of both.
> 
> Didn't you and I have a Zoom call a few months ago where we agreed that those shouldn't be mixed, or am I misremembering that (in that case, sorry)?

@peacekeeper Personally, I believe SD-JWT VC is the simpler format and there is no need for the `vcdm` object. However, there are people that have defined data models based on VCDM and want to use it with SD-JWT VC, hence this proposal. Segregating this based on a dedicated `vcdm` claim does make sense to me to avoid issues with conflating layers (hence the suggested processing model https://github.com/openid/oid4vc-haip/pull/147#discussion_r1922554997) and with JSON-LD for the reasons outlined in here https://github.com/openid/oid4vc-haip/pull/147#discussion_r1922712373.

Reviewer 8: @awoie It´s not the complexity it`s the fact that use cases where ypou need to integrate external schemes are not possible with SD_JWT VC. Same with microcredentials. So SD_JWT VC is easy but critical for certain use cases, similar if you want to use DID as done in industry use cases

Reviewer 5: @steffenschwalm I guess then we agree :)

Reviewer: > there are people that have defined data models based on VCDM and want to use it with SD-JWT VC

Why do they want that? Maybe because they have been misled by statements such as "selective disclosure with approved crypto is only possible with SD-JWT" or similar?

Is there any real technical justification why SD-JWT VCDM is needed or useful, or is this basically an exercise to satisfy some obscure political interests?

Reviewer 8: @peacekeeper see my comment. The easiee selective Disclosure is not the only reason. It's no political reason. Would recommend to focus on technical subjects. If you want political discussions let's do it at a Bar in future

Reviewer: Okay let's focus on technical subjects. I still haven't heard a single valid technical argument for this work. Here is what I have heard so far:

- "SD-JWT VCDM is needed because some use cases don't want the complex JSON-LD stuff." -> Fine, but then just use JWT. Or SD-JWT. Or SD-JWT VC. No SD-JWT VCDM needed.
- "SD-JWT VCDM is needed because some use cases like microcredentials need JSON-LD. -> Fine, but then just use VCDM. No SD-JWT VCDM needed.
- "We need this to do selective disclosure with VCDM." -> Not a valid argument, because VCDM supports selective disclosure with Data Integrity, using simple crypto (like SD-JWT) or advanced crypto (BBS+, etc.).
- "We need this to do SIMPLE selective disclosure with VCDM." -> Not a valid argument, because you can already do VCDM with SD-JWT, no SD-JWT VCDM needed.
- "We need this to use the same selective disclosure mechanism for all credentials." -> Not a valid argument, because you can already do VCDM with SD-JWT, no SD-JWT VCDM needed.

So please, WHY is this needed?

The only argument in this thread so far that's left is a NON-technical argument: "There are people that want it".

Reviewer: > JWT does not solve requirements.

I guess I just don't know what the requirements are that make this SD-JWT VCDM work necessary and can't already be fulfilled without it. I can't find such a requirement anywhere in this PR content or comments. If the requirement is to create an acronym that gives the illusion of unifying two incompatible data models, then I agree SD-JWT VCDM meets this requirement.

Reviewer 6: @peacekeeper What other ways are there to do selective disclosure with data integrity (besides SD-JWT, ZKP)?

Reviewer: @danielfett There is ecdsa-sd-2023, which is similar to SD-JWT (supporting selective disclosure using simple crypto), but aligned with Data Integrity:
- https://www.w3.org/TR/vc-di-ecdsa/#ecdsa-sd-2023
- https://github.com/digitalbazaar/ecdsa-sd-2023-cryptosuite

Reviewer 5: @danielfett I believe @peacekeeper is referring to https://www.w3.org/TR/vc-di-ecdsa/#ecdsa-sd-2023 which is an SD mechanism based on the data integrity specification. I raised some concerns on this approach here https://lists.w3.org/Archives/Public/public-vc-wg/2023Aug/0035.html. There are probably other concerns especially wrt data integrity itself as well but this is not the right forum to discuss pros and cons.

Author: the main point here was sd-jwt vc compliance while reusing vcdm 1.1/2.0, so i am ok dropping this last part.

```suggestion
SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].
```

Reviewer 8: @Sakurann would keep the original one as this was exactly the idea as you mentioned and would allow to use the exact same selective disclosure mechanism for all credentials as Daniel mentioned.

Reviewer 2: Agree with @steffenschwalm here. There is also SD-CWT, ISO mStuff, vc-xose, and probably others I don't know about that can do selective disclosure with various tradeoffs and preferences. None of that makes the "while enabling selective disclosure" here untrue.

Reviewer 7: Is there a better description of:
SD-JWT VCDM is a credential format that is compliant to IETF SD-JWT VC [@!I-D.ietf-oauth-sd-jwt-vc] that allows the usage of existing data structures represented using [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0].

Reviewer 7: @danielfett A way of realizing selective disclosure is using bbs+ signatures, selective claim signatures and presentations possible with W3C VC DM. Isn‘t SD-JWT bound to JWS?

Reviewer 7: > What other ways are there to do selective disclosure with data integrity (besides SD-JWT, ZKP)?

One way of doing SD besides SD-JWT is to implement BBS+ signatures to enable selective disclosure of Verifiable Credential (VC) claims and prepare for Zero-Knowledge Proof (ZKP) integration using schemes like Groth16. Below is an overview:

1. Credential Payload Construction
The payload includes `credentialSubject` and metadata, ensuring it supports selective disclosure.

`payload = {
    "iat": int(datetime.now(tz=timezone.utc).timestamp()) - 60,
    "iss": issuer_did,
    "jti": uniqID,
    "vc": {
        "@context": ["https://www.w3.org/2018/credentials/v1"],
        "type": ["VerifiableCredential", "StudentIDCard"],
        "credentialSubject": credential_subject,
    },
}
`

2. Flattening and Signing Claims with BBS+
Claims are flattened, enabling selective disclosure. BBS+ signs each claim individually.

`flattened_payload = flatten(payload, '.')
to_sign = [json.dumps({key: flattened_payload[key]}, ensure_ascii=False) for key in sorted(flattened_payload.keys())]
signer = bbs_core.SignRequest(to_sign, bbs_dpk, bbs_secret)
signature_bytes = base64.b64encode(signer.sign_messages().signature).decode()
`

3. Returning the Signed Credential
The signed payload is packaged into a JWT, and a nonce is included for verifier challenges.

`vc_jwt = jwt.encode(payload, private_key, algorithm="ES256", headers={"kid": issuer_kid})
return jsonify({"credential": vc_jwt, "signature": signature_bytes, "c_nonce": generate_nonce(10)}), 200
`

4. ZKP Integration for Advanced Privacy
For enhanced privacy, the flattened claims and BBS+ signatures can be extended to support Groth16 proofs, allowing holders to prove knowledge of signed claims without revealing them. This enables use cases such as anonymous attribute verification.

`# Example Groth16 pseudocode:
proof = groth16.generate_proof(credential_claims, bbs_signature, zk_snark_keys)
verifier.verify_proof(proof, public_inputs)` (not implemented yet)

Differences from SD-JWT VC Data Model
1. BBS+ Signatures for Selective Disclosure:
Unlike SD-JWT VC, which uses a salted hash approach for selective disclosure, we integrate BBS+ signatures. BBS+ signatures enable cryptographically efficient selective disclosure and unlinkability without revealing other claims.
2. ZKP Integration:
Our system is designed to support Zero-Knowledge Proofs (e.g., Groth16), which allow holders to prove knowledge of claims without directly revealing them. This enhances privacy beyond the standard SD-JWT VC model.
3. Credential Type and Semantics:
While SD-JWT VC uses "Type Metadata" to define the type/schema of credentials, our implementation incorporates the W3C Verifiable Credentials `@context` for RDF-based semantic richness and interoperability. (please correct me if this is not correct)
4. Additional JWT Extensions:
We include custom claims such as `bbs_dpk` (deterministic public key) and `nonce` for verifier challenges, enabling advanced cryptographic use cases not directly covered by SD-JWT VC.

Our code is written in python (soon be publicly available) and uses a Rust bbs+ lib.
Lessons learned, next time, we start directly with Rust :-)

Reviewer 7: Can someone briefly summarize why one should use SD-JWT VCDM instead of W3C VC DM? As we were asked to participate here, I still don't understand why one should use a Golf 1 instead of a McLaren f1. We have implemented several different issuers over the past years, and the only reasonable data format for issuing, presenting, and verifying VCs is based on the W3C VC DM model. So, what are the technical arguments?

Reviewer 2: > As we were asked to participate here:

@pherbke, would you mind saying who "we" is and who asked them to participate here and why? That might help others here understand how best to try and answer your questions.

Reviewer 7: @bc-pi That was a misunderstanding and I deleted my questions already. Sorry for that.

Author: ok, keeping as-is as a starting point.

SD-JWT VCDM credentials contain a data structure that can be processed using a JSON-LD processor after applying SD-JWT processing.

When IETF SD-JWT VC is mentioned in this specification, SD-JWT VCDM defined in this section MAY be used.

For backward compatibility with JWT processors, the following registered JWT claims MUST be used, instead of their respective counterpart properties in [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0]:

> * `exp` Claim MUST represent the `expirationDate` property, encoded as a UNIX timestamp (NumericDate).

Reviewer 5: I'd suggest removing the requirement on `exp`. This is why I landed there:

- **SD-JWT VC**: Validity claims are **security-critical**; for example, a verifier must reject an expired SD-JWT VC. While use cases can define additional business-logic-specific validity claims, many use cases likely won’t require this and domain experts/use case-specific data model authors might reuse SD-JWT VC validity claims, i.e., `iat`/`nbf`/`exp` but this definition is out of scope of SD-JWT VC/VCDM.

- **VCDM**: Validity claims (e.g., `issuanceDate`/`expirationDate` in VCDM 1.1 or `validFrom`/`validUntil` in VCDM 2.0) are **not security-critical**. Instead, they pertain to information associated with the `credentialSubject` and are part of the business logic. For example, the business logic can decide to accept an expired VCDM for specific use cases.

- **ISO mdoc**: Validity claims in ISO mdocs (e.g., the `ValidityInfo` structure) are **security-critical** and integral to the verification algorithm. For instance, a verifier must reject an expired Mobile Security Object (MSO). However, use cases can define additional claims specific to the business logic, such as `expiry_date` for an mDL.

The following is out of scope of  SD-JWT VC/VCDM but are legitimate scenarios:
- For SD-JWT VCDM **business logic processors**, it can be acceptable for `exp` and `expirationDate` to differ in some cases. The VCDM business logic receives the verified credential and applies additional validation rules on `expirationDate` as needed. For instance, an expired `expirationDate` may still be usable for certain purposes, as defined by domain experts or data model authors for a specific use case.
- On the other hand, for example, domain experts or data model authors of a use case might choose not to use the `vcdm` claim at all. Instead, they may define that business-logic-specific validity info is also expressed by `iat`, `nbf`, or `exp`.

I hope this makes sense.
 
```suggestion
```

Reviewer 4: most popular use case:
- token can be expired
- it is still valid that I'm over 18
Over 18 business verification could accept expired tokens (e.g., offline, no chance to re-fill) for over 18, as the claims are still true.

It might never happen, but I agree that exp and validity are two different elements;
More than anything, it is crucial to clearly define how to express validity.

Reviewer: Distinguishing between "security-critical validity" and "business-logic-specific validity" sounds very fuzzy, but I could see how expiration of a digital token and expiration of a VC document may be two different things that could be modeled with two different timestamps.

On the other hand, I recall that the VC community ran into LOTS of interop problems the [last time this kind of JWT / VCDM hybrid was attempted](https://www.w3.org/TR/vc-data-model/#json-web-token), e.g. see issue https://github.com/w3c/vc-data-model/issues/809 and many others.

Reviewer 5: > On the other hand, I recall that the VC community ran into LOTS of interop problems the [last time this kind of JWT / VCDM hybrid was attempted](https://www.w3.org/TR/vc-data-model/#json-web-token), e.g. see issue [w3c/vc-data-model#809](https://github.com/w3c/vc-data-model/issues/809) and many others.

If I remember correctly, these interop issues happened because of the transformation logic, e.g., `iat` <-> `issuanceDate` which had different types (UNIX timestamp vs ISO date time) and which had an optimisation to omit certain VCDM claims if the JWT claims were used. The proposal here is that `iat`/`exp`/`nbf` have to be used for signature lifetime (and for non-`vcdm` VCs, it could be used for document lifetime as well) and domain-specific claims or VCDM claims (if `vcdm` claim is present) can be used for document lifetime.

Author: ok, accepted text along the lines of what Oliver has suggested.

> * `iss` Claim MUST represent the `issuer` property.

Reviewer 5: I would keep the original line for the following reason.

The SD-JWT VC processing is first since it deals with security-critical functions. `iss` is used by SD-JWT VC to validate the public verification key of the issuer (as per SD-JWT VC). To make sure self-declared und unverifiable values are not permitted, the linkage between the `iss` and `vcdm.issuer` (if String) /`vcdm.issuer.id` (if `vcdm.issuer` is an object) has to be kept for integrity reasons since the business logic (if `vcdm` claim is included) would rely on the SD-JWT VC verification logic that also checks the issuer.

> * `iat` Claim MUST represent `issuanceDate` property, encoded as a UNIX timestamp (NumericDate).

Reviewer 5: I'd remove the restriction on `iss` for the same reason as above https://github.com/openid/oid4vc-haip/pull/147/files#r1922646896
```suggestion
* `iat` Claim MUST represent `issuanceDate` property, encoded as a UNIX timestamp (NumericDate).
```

Reviewer 4: In eIDAS official timestamps are used to express time of issuance, whenever necessary. iat claim is defined as self-proclaimed time of issuance.
IMO it should be defined
a) when iss should be used and how (e.g., time of signing)
b) when timestamps should be used
Also same comment:
- time of signature creation
- time of issuance (e.g., diploma issuance date is different from the time of signing)

Reviewer 9: agree to Alen, when timestamp should be used the use of RFC 3161 timestamps shall be possible

Reviewer 5: IMO, `iat` is the time of issuance asserted by the signer which is normally also the issuer, i.e., when the signature was produced. I understand that in eIDAS terms, this might be different. 

@alenhorvat imo, time of issuance date of the diploma is out of scope of this completely. This relates to the data model for a specific domain/use case. If domain experts have to define this, they could use either 1) `iat` if both are the same (most simple), b) use `iat` to express for time of VC creation and something else, e.g., `issuanceDate`/`validFrom` (as per W3C VCDM), or 3) even `myDomainSpecificDiplomaIssuanceDate` if they wanted to.

Reviewer 8: @awoie time of VC issuance and time of signature creation is practically not necessarily same especially not in context of eiDAS for which SD-JWT VCDM is purposed. and  encoded as a UNIX timestamp (NumericDate) does not match to RFC 3161 where it´s timestamp token typically ASN1 encoded. Assume that @alenhorvat refers to this, recommend to change

Author: accepting the text along the lines of what Oliver suggested above https://github.com/openid/oid4vc-haip/pull/147#discussion_r1922678753

[ARF](https://eu-digital-identity-wallet.github.io/eudi-doc-architecture-and-reference-framework/1.4.0/annexes/annex-3/annex-3.01-pid-rulebook/) defined `issuance_date` as Date (and possibly time) when the PID was issued, so it is necessary to document time of signature. iat claim is [defined](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6) as NumericDate. if you need timestamp on top please use json serialization etc.

Reviewer 4: Issuance date and cryptographic timestamp are two different things. Timestamps can be used with both compact and JSON serialisation as defined in ETSI (JAdES; see baseline-B for compact serialisation and any of the remaining three for JSON serialisation); timestamp is following the RFC 3161 (https://www.ietf.org/rfc/rfc3161.txt). Depending on the profile, different parts of the content/header/a combination of are timestamped.
Timestamp is defined by the signature profile and are created during signature creation.

issuance date/valid from is information related to the content and is created at the time of content creation.

As implementing acts are published, let's please refer to the data model in the implementing act:
https://eur-lex.europa.eu/eli/reg_impl/2024/2977/oj

The implementing regulation solves the problem: there's no mention of issuance date and that's left to the signature profile.

Reviewer 8: Fully agree to @alenhorvat  as ARF is not legally mandatory in Europe @Sakurann. The spec shall fulfill the law

* `status` Claim MUST represent `credentialStatus` property.
* `sub` Claim MUST represent the `id` property of `credentialSubject` property.

IETF SD-JWT VC is extended with the following claims:

> * `vcdm`: OPTIONAL. Contains properties defined in [@!W3C.VCDM1.1] or [@!W3C.VCDM2.0] that are not represented by their counterpart JWT Claims as defined above.

Reviewer: How is `"vcdm"` different from the good old [`"vc"` claim](https://www.w3.org/TR/vc-data-model/#iana-considerations)?

Reviewer 4: I recommend removing the "vcdm" if the governance of the data model is well defined. (see my question on how to express claims from different vocabularies)

Reviewer 5: The `vcdm` claim is there to allow JSON-LD processors to process the `vcdm` object which has to be a valid JSON-LD object in compact form. Some JSON-LD processors would have issues with terms that are not included in a context definition. There might be some SD-JWT and SD-JWT VC claims today or in the future that are not included and therefore cause issues. By encapsulating the JSON-LD stuff within `vcdm`, we guarantee this won't cause any issues.

Reviewer: > By encapsulating the JSON-LD stuff within `vcdm`

Isn't that exactly the same reason why you introduced `vc` a few years ago? Why not re-use that?

Reviewer 6: I guess `vc` would be fine, but it is an IANA-registered claim that's defined for use in a different context. Might be okay to re-use though. `vcdm` would make it clear that this claim is to be used with the processing model in mind that @awoie describes in one of the other suggested changes.

Author: - it does not repeat the errors from the past because vcdm is much better defined in the context of sd-jwt vcdm. 
- it could be `vc` claim, personally think `vcdm` is cleaner because IANA reserved `vc` for w3c vcdm 1.1. 
- we could probably do without `vcdm` at all, but processing using different processors is slightly easier when `vcdm` claim is present, so would prefer to keep it. 

please re-read the above paragraphs on the purpose of sd-jwt vcdm - what you propose in your third bullet point there is not what sd-jwt vcdm is for; for that, please just use w3c vcdm.

Reviewer 4: VCDM should not be there as there's no need for it if we follow the processing logic above (which worked well with the existing use cases, creating a clear boundary and positioning the JSON-LD so that it can be used whenever needed)

Reviewer: > * IANA reserved `vc` for w3c vcdm 1.1.

This raises the question whether `vcdm` is  only for VCDM 1.1 or for both VCDM 1.1 and VCDM 2.0 ?

Author: vcdm is intended for both VCDM 1.1 and VCDM 2.0 - I thought text made it pretty clear?

Reviewer 4: 'vcdm' should not be there as there's no need for it if we follow the processing logic above (which worked well with the existing use cases, creating a clear boundary and positioning the JSON-LD so that it can be used whenever needed)

Reviewer 5: I think `vc` won't be used. It would be `"vcdm": { VCDM 1.1 or VCDM 2.0 stuff }`.

Reviewer 4: Why you need that?

See:
- base data format is JSON
- if use cases want to use JSON-LD, JSON-LD should cover all the definitions needed to successfully process the data model (again, depending on for what purpose JSON-LD is used)
- JSON-LD context MUST be treated as any referenced content in the data model (integrity protection, ability to share the data as unsigned payload);

Why this cannot work?

Reviewer 5: Perhaps this is my lack of real world JSON-LD processors but wouldn't this require that all things that are not part of the JSON-LD VCDM context have to be defined in proper JSON-LD context to not cause issues with these JSON-LD processors, e.g., `status`, `vct`?

Reviewer 4: Yes, and the proposal is that if JSON-LD is used, the JSON-LD context must cover all the claims. It leads to a very clean design without exception handling.

Reviewer 4: If implementations had issues
a) what are the issues?
b) can we contact them and learn more about the issues?

Reviewer: > I think `vc` won't be used. It would be `"vcdm": { VCDM 1.1 or VCDM 2.0 stuff }`.

But @Sakurann said:

> IANA reserved `vc` for w3c vcdm 1.1

Wouldn't there be a lot of ambiguity and interop problems if you define two different JWT claims that can be used for the same purpose (VCDM 1.1)?

Reviewer 6: I don't see how that would be a problem. The SD-JWT VC DM wrapper is new and has to be implemented anyway; any claim name would work. What is important is what is in the `vcdm` claim.

Reviewer: > any claim name would work

So are you saying if an Issuer wants to use VCDM 1.1 wrapped inside a JWT, they could use either the `vc` or `vcdm` claim, since both serve the same purpose? And would Verifiers also be expected to support both? Or pick one of the two?

Reviewer 6: I meant to say that we can define any claim name here, without much impact on implementations. Whether they process `vcdm`, `vc` or `foo` doesn't matter. There may be a preference of one over the other for historic reasons, or because one has an existing IANA definition, or because one has clearer semantics, but other than that, it doesn't matter.

> 

Reviewer 5: ```suggestion
The following outlines a suggested non-normative processing model for SD-JWT VCDM:

1. SD-JWT VC Processing:
- A receiver (holder or verifier) of an SD-JWT VCDM applies the processing rules outlined in SD-JWT VC Section 4 of [@!I-D.ietf-oauth-sd-jwt-vc], including verifying signatures, validity periods, status information etc.
- If the `vct` value is associated with any SD-JWT VC Type Metadata, schema validation of the entire SD-JWT VCDM is performed, including the nested `vcdm` claim.
- Additionally, trust framework rules are applied, such as ensuring the issuer is authorized to issue SD-JWT VCDMs for the specified `vct` value.
2. VCDM Processing (business logic):
- Once the SD-JWT VCDM is verified and trusted by the SD-JWT VC processor, the receiver extracts the VCDM (1.1 or 2.0) object from the vcdm claim.
- The extracted VCDM is then passed to the business logic for further use case-specific processing and validation. The business logic assumes that all security-critical functions (e.g., signature verification, trusted issuer) have already been performed during the previous step.

```

Reviewer 5: @peacekeeper Thanks, that was a typo. I updated the suggestion above to make this more clear.

The following is a non-normative example of an unsecured payload of an SD-JWT VCDM, that is built using the example of unsecured payload in Section 3.3 of [@!I-D.ietf-oauth-sd-jwt-vc]:

```json
{
  "vct": "https://credentials.example.com/identity_credential",
  //W3C VCDM 2.0 compliant claims
  "vcdm": {
    "@context": ["https://www.w3.org/ns/credentials/v2"],

>     "type": ["VerifiableCredential", "https://credentials.example.com/identity_credential"],

Reviewer 5: I would remove the relationship between `vct` and `type`. We should acknowledge that the typing can be different.  I assume that JSON-LD processors want their JSON-LD type system. To guarantee interop on the protocol level, e.g., query credentials in OID4VP, the `vct` values has to be used.
```suggestion
    "type": ["VerifiableCredential", "IdentityCredential"],
```

Reviewer: This means you could have a situation where you request a driver's license on the OID4VP level, but then receive a VC with a type of birth certificate?

Reviewer: > To guarantee interop on the protocol level, e.g., query credentials in OID4VP, the vct value has to be used.

I always thought OID4VP supports different credential formats? Is it now getting more tightly coupled to SD-JWT VC?

Reviewer 5: > This means you could have a situation where you request a driver's license on the OID4VP level, but then receive a VC with a type of birth certificate?
@peacekeeper It is the issuer's responsibility to avoid nonsensical situations like this. If the issuer is trusted, I'd assume this not going to happen and they would make sure that `vct` value and business logic object would align. For instance, this could be done via `vct` Type Metadata where the Type Metadata would say that for a given `vct` value the `type` property has to have to certain value as well.

Reviewer 5: >I always thought OID4VP supports different credential formats? Is it now getting more tightly coupled to SD-JWT VC?
@peacekeeper It is not getting more tightly coupled. OID4VP does support any format. This PR is specifically about SD-JWT VCDM. Also note that this is the HAIP repository, not the OID4VP repository.

Reviewer: I don't understand the point of a hybrid VC format that allows two different places in the data model where types can be expressed using two different mechanisms. This clearly creates huge potential for inconsistencies and non-interoperability.

Reviewer 6: The goal is to have the selective disclosure capabilities of SD-JWT VC, but re-use existing data structures in W3C VCDM format. See the modified intro text: "SD-JWT VCDM credentials contain a data structure that can be processed using a JSON-LD processor after applying SD-JWT processing."

Reviewer: > The goal is to have the selective disclosure capabilities of SD-JWT VC, but re-use existing data structures in W3C VCDM format.

Apart from the fact that VCDM already supports selective disclosure using Data Integrity proofs (no **SD-JWT** needed), it ALSO support securing VCDM with SD-JWT directly (no **SD-JWT VC** needed).

So why SD JWT VCDM?

Reviewer 5: > So why SD JWT VCDM

If you asked me personally, one major advantage is that no JSON-LD is required whereas VCDM 2.0 requires the payload to be a valid JSON-LD object in compact form. It has more overhead and the idea is that while this should stay possible, it shouldn't be required. SD-JWT VCDM would allow for this.

Reviewer 7: When discussing technical subjects, a question:
How can ZKPs and future signature technologies like lattice-based signature schemes (e.g., Dilithium) be integrated into the SD-JWT VC (DM)? Is there a point in eidas2.0? Is this even important?

Reviewer 4: - base data format is JSON
- if use cases want to use JSON-LD, JSON-LD should cover all the definitions needed to successfully process the data model (again, depending on for what purpose JSON-LD is used)
- JSON-LD context MUST be treated as any referenced content in the data model (integrity protection, ability to share the data as unsigned payload); See related discussion: https://github.com/w3c/vc-data-integrity/issues/323#issue-2717232488

Hence, I propose to remove the "VCDM" claim and the double-processing rules as it is creating confusion and repeating the error from the past.

    "credentialSubject": {
      "given_name": "John",
      "family_name": "Doe",
      "email": "johndoe@example.com",
      "phone_number": "+1-202-555-0101",
      "address": {
        "street_address": "123 Main St",
        "locality": "Anytown",
        "region": "Anystate",
        "country": "US"
      },
      "birthdate": "1940-01-01",
      "is_over_18": true,
      "is_over_21": true,
      "is_over_65": true
    }
  }
}
```

The following is a non-normative example of how an unsecured payload of an SD-JWT VCDM above can be used for the SD-JWT:

```json
{
  "_sd": [
    "vwUhdFvpylx9Sqi2YNBV1dfVK6lCbhXvkH0nThfKFT0"
  ],
  "iss": "https://example.com/issuer",
  "iat": 1683000000,
  "exp": 1883000000,
  "vct": "https://credentials.example.com/identity_credential",
  "_sd_alg": "sha-256",
  "cnf": {
    "jwk": {
      "kty": "EC",
      "crv": "P-256",
      "x": "TCAER19Zvu3OHF4j4W4vfSVoHIP1ILilDls7vCeGemc",
      "y": "ZxjiWWbZMQGHVWKVQ4hbSIirsVfuecCE6t4jT9F2HZQ"
    }
  }
}
```

> # Crypto Suites

Reviewer 4: Is anyone involved with the work on PQC signatures?

Reviewer 7: @Alen: Do you mean post-quantum crypto?

Author: can you please open a separate issue in HAIP? :)
